### ISSUE TITLE
Sort by tag

### ISSUE THREAD

**Issue by ghost** [emacs-elfeed/elfeed#203]
Is there a way to sort my feeds by tag? I'm not saying look at a specific tag but in my entire list of feeds order them by there tag alphabetically.

**Comment by skeeto** [comment]
There is currently no way to do this. However I can imagine a potential
feature that allows users to supply a custom sort function with the
understanding that it comes at a performance cost — the database is
indexed only by time.

**Comment by ra1nb0w** [comment]
Should be nice to have a way to group by tag. I mean, have feeds with the same tag one after another. Could _elfeed-search-sort-function_ be used for this scope?
thanks

**Comment by skeeto** [comment]
@ra1nb0w Yes, `elfeed-search-sort-function` should do it, with all the caveats of using a custom sort. This feature was added less than a month ago (b62a2d0040f1a8ad0a23a48e3843d2ed5cbc2a8c) and I forgot about it applying to this issue.

**Comment by ra1nb0w** [comment]
with all the caveats what do you mean except break the live search?

**Comment by skeeto** [comment]
I should have said "caveat," singular. It just breaks the live search. 
Live search needs to be fast, and it relies on the time index to quickly 
fill the visible part of the buffer with results. Under an alternative 
sort, all that's available is a very slow "table scan," though it 
doesn't actually fall back on this.

During live search it picks out just enough matching entries to fill the 
display, starting with newest first, then applies the custom sort. Under 
correct (non-live) operation, the custom sort might list a very old 
entry first, but in live search this entry wouldn't be included at all 
due to that initial selection process.

**Comment by ra1nb0w** [comment]
ok. Thanks for the explaination.

**Comment by gregf** [comment]
@skeeto any chance you could show me an example on how I would sort by tag with this new function? Thanks.

**Comment by skeeto** [comment]
This example serializes the tag list into a string (simple, but 
inefficient), then sorts entries based on that string, falling back on 
the date to break ties (descending).

    (defun my-elfeed-tag-sort (a b)
      (let* ((a-tags (format "%s" (elfeed-entry-tags a)))
             (b-tags (format "%s" (elfeed-entry-tags b))))
        (if (string= a-tags b-tags)
            (< (elfeed-entry-date b) (elfeed-entry-date a)))
          (string< a-tags b-tags)))

    (setf elfeed-search-sort-function #'my-elfeed-tag-sort)

Maybe my tags are too much of a mess for this to work, but I didn't find 
this tag sort to be useful.

**Comment by catweazle9** [comment]
I think the tags comparison should be in the 'else' clause of the date comparison, i.e.:
      (if (string= a-tags b-tags)
          (< (elfeed-entry-date b) (elfeed-entry-date a))
        (string< a-tags b-tags))))

**Comment by minad** [comment]
In Emacs 30 one can use value< for comparison:

~~~ elisp
(defun my-elfeed-tag-sort (a b)
  (let* ((a-tags (elfeed-entry-tags a))
         (b-tags (elfeed-entry-tags b)))
    (if (equal a-tags b-tags)
        (< (elfeed-entry-date b) (elfeed-entry-date a)))
    (value< a-tags b-tags)))

(setf elfeed-search-sort-function #'my-elfeed-tag-sort)
~~~
